src/models/room.ts

```diff
@@ -940,12 +955,42 @@ export class Room extends ReadReceipt<RoomEmittedEvents, RoomEventHandlerMap> {
         if (nonFunctionalMemberCount > 2) return;
 
         // Prefer the list of heroes, if present. It should only include the single other user in the DM.
-        const nonFunctionalHeroes = this.summaryHeroes?.filter((h) => !functionalMembers.includes(h));
+        const nonFunctionalHeroes = this.heroes?.filter((h) => !functionalMembers.includes(h.userId));
         const hasHeroes = Array.isArray(nonFunctionalHeroes) && nonFunctionalHeroes.length;
         if (hasHeroes) {
+            // use first hero which exists
+            for (const hero of nonFunctionalHeroes) {
+                if (!hero.displayName && !hero.avatarUrl) {
+                    // attempt to look up renderable fields from the m.room.member event if it exists
```

Comment:
why do we check if the user is a member of the room in this case, but not in the case that `hero.displayName` or `hero.avatarUrl` have been set?

Reply:
I asked Kegan and it sounds is intentional if the server doesn't populate either field, it will essentially fall back to the old behaviour:

> If we have a Hero then the server has explicitly sent data back and we should use it. But, in the case of where there is no actual data for this hero (avatar/display name) then rather than send nothing we fallback to the m.room.member event's displayname/avatar_url if they exist, as it might have some data, which is better than displaying nothing. So effectively the codepath for v2 fallback because-there-is-no-Hero and SSS-Hero-has-no-data is the same

I've tweaked the comment a bit to reflect this.

Reply:
Oh, as in should it still ensure that the user is definitely in the room even if all the information comes from the hero object?